### PR TITLE
Add start of tutorial to run flux on k8s

### PR DESCRIPTION
**Pull request by vsoch**
This is a refactor of the 2022 tutorial that will use the helm chart for jupyter + our custom containers to spawn flux pods on k8s. This is mostly working on GCP and needs to be tested on EKS, specifically for basic functionality and then for the SSL stuffs. I don't expect this to be fully working for 2023 so set expectations low, but hey, you never know!  A few quick notes:

- we use helm charts to deploy 
- the containers have been updated for k8s (based on their default bases)
- the entire tutorial contents are provided via their own repository (in the Dockerfile) as a git volume. Anything added to home doesn't actually stick and this is the suggested way.
- The entrypoint had some issues, more detail below.

I was having trouble getting flux start to take, and it took me like 15 times to realize this, but the CMD provided to the container is useless. The reason is because it's written into the helm chart. I started tweaking that, but it didn't actually work wtth tini to just give it a flux start - even when I provided full paths to flux / jupyter tini was like "beep boop no such file or directory" so you have to give it an entrypoint explicitly:

```yaml
# This is the "spawn" image
singleuser:
  image:
    name: vanessa/flux-jupyter-spawn
    tag: latest
    pullPolicy: Always
  cpu:
    limit: 1
  memory:
    limit: '4G'
  # This apparently is given to tini, either in the yaml or via our entrypoint
  # Things get weird when you remove tini so I tried to keep it
  cmd: /entrypoint.sh
```

and then in the entrypoint the environment seems preserved, and then the entire thing starts under flux start. 

```bash
#!/bin/bash

/usr/bin/flux start --test-size=4 /usr/local/bin/jupyterhub-singleuser
```
And then hey - we have working flux!

image

So it wasn't starting before because it wasn't even using the command from the container. 

This is mostly working now as in, it's in the state I remember it where the login page works, the console/interface works, and flux works! On GCP we'd just get some domain for this proxy, but I suspect on AWS it's different. Consider this a WIP because I need to move it over to AWS to test, and then look at the SSL bits.

### CODE REVIEW

**Comment by vsoch** [comment]
Update: aws config is added. I learned that the gitRepo volume type is deprecated, but an init container with an empty volume mounted to where we need to have joyvan's home (shared by the container) seemed to resolve that issue. So now both are working the same, and next steps are to add the SSL or https. I put a commented out section in the config-aws.yaml. And the steps are to:

 - generate the managed certificate in the aws console, that should give you an ARN
 - add the arn and an openssl generated secret to the same file
 - trying deploying with that bit uncommented and hope it works?

**Comment by milroy** [comment]
We can merge this once you include the updates from our hackathon.

**Comment by vsoch** [comment]
Summary updates:

- Tutorial files reorganized and updated
- storage for spawner set to none
- min/max added and tested (needs more testing!)
 
And Details:

- I reorganized files and the README, adding more links / details, and ensuring each content type is in a proper subdirectory (I don't like it when everything is sploot out in the root with a gazillion different files).
- I tested storage, and removing storage all together I think is our best bet. I confirmed that when we have the type set, each deployed pod creates a volume. When we don't, there is only one for the hub (ideal).

```console
$ kubectl get pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
hub-db-dir   Bound    pvc-d8981bc2-733a-40f7-b577-0aacfb8fb6ff   1Gi        RWO            gp2            5m36s

$ kubectl get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                STORAGECLASS   REASON   AGE
pvc-d8981bc2-733a-40f7-b577-0aacfb8fb6ff   1Gi        RWO            Delete           Bound    default/hub-db-dir   gp2                     5m34s
```

- I tried refactoring so the tutorial files / notebook are all here, but then couldn't get the right init container command to clone and move to the subdirectory. Setting a different working directory didn't take. So I fell back to using the original (external) repo and we can likely improve upon this in the future (or just make branches over there). https://github.com/rse-ops/flux-radiuss-tutorial-2023
- I re-did the local tutorial instructions - the old way didn't work with the new container. We need an entrypoint with ip set to 0.0.0.0.
- I updated the tutorial / files, removing "mini," adding more references, and a new flux watch example. For the others from Mark's email, I either couldn't find an example in flux core, or it seemed like an advanced use case not appropriate for here, or it was more something that would go in slides. This was my judgment of course and might be wrong.
- I updated the GCP config.yaml to use the same strategy, although this is not tested. I'm OK testing later / when we need.
- I tested auto scaling (details below)

For setting the min/max size.  When I set desired to 1, min to 1, and max to 3, it started with 3 nodes, but never increases.

```bash
$ kubecl get nodes
NAME                                            STATUS   ROLES    AGE     VERSION
ip-192-168-105-191.us-east-2.compute.internal   Ready    <none>   6m28s   v1.24.15-eks-a5565ad
ip-192-168-135-46.us-east-2.compute.internal    Ready    <none>   6m27s   v1.24.15-eks-a5565ad
ip-192-168-160-5.us-east-2.compute.internal     Ready    <none>   5m45s   v1.24.15-eks-a5565ad
```

I verified there was no activity in the console. I think we need to add this (but I didn't try yet because would require bringing up/down an entire new cluster, not sure I have time tonight). 

```
    #withAddonPolicies:
    # autoScaler: true
```
See here: https://eksctl.io/usage/autoscaling/ I also wanted to share another spurious new error - 1/5 times I saw the pull of a pause container timed out and the entire thing failed:

```
  Warning  Failed     28s                  kubelet            Failed to pull image "registry.k8s.io/pause:3.9": rpc error: code = Unknown desc = failed to pull and unpack image "registry.k8s.io/pause:3.9": failed to resolve reference "registry.k8s.io/pause:3.9": failed to do request: Head "https://us-east4-docker.pkg.dev/v2/k8s-artifacts-prod/images/pause/manifests/3.9": read tcp 192.168.105.191:41010->142.251.176.82:443: read: connection timed out
  Warning  Failed     28s                  kubelet            Error: ErrImagePull
```
So I had to start over. It worked the second time. 

## Summary

High level - on tutorial day make sure there is extra time to spin things up! The remaining TODO is that we need to test adding the autoscaling true / otherwise adding a policy so it works. If it doesn't we will need to manually do it. Just adding the min/max and spawn limit doesn't do it because there are no policies (and likely eksctl needs to make them?) So that should be set higher and tested again! @milroy when this looks good to you it should be good to merge, and I can probably spin this up tomorrow and give the scaling one more shot.

File: 2023-RADIUSS-AWS/JupyterNotebook/login.html [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+{% block login %}
+<div id="login-main" class="container">
+{% block login_container %}
+
+  <div class='container branding'>
+    <a href="https://flux-framework.org">
+      <img src="https://flux-framework.org/assets/images/Flux-logo.svg">
+    </a>
+      <div style="text-align: center; padding-bottom:30px">
+        <h1 class="text-normal" role="banner" align="center">RADUISS tutorial running on AWS</h1>
```
[the remark this answers is not in the document]
Comment (vsoch) [reply 2023-RADIUSS-AWS/JupyterNotebook/login.html]: Ha! I didn't even notice."RAD-OO-EESE" sounds much more exotic.

File: 2023-RADIUSS-AWS/JupyterNotebook/aws/eksctl-config.yaml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+              - "ec2:DescribeVolumes"
+              - "ec2:DetachVolume"
+            Resource: '*'
+
+
+managedNodeGroups:
+  - name: ng-us-east-2a
+    instanceType: m5.large
+    volumeSize: 30
+    desiredCapacity: 1
```
[the remark this answers is not in the document]
Comment (vsoch) [reply 2023-RADIUSS-AWS/JupyterNotebook/aws/eksctl-config.yaml:64]: Tested - see note below. TLDR it's not enough and I think we need to add a policy (if we want it automated).

**Comment by milroy** [comment]
That's a big addition, all of which sounds great. I'll take a look in the morning and wait for your scaling test before I merge this.

**Comment by milroy** [comment]
I took the liberty of dropping [WIP] since we'll merge this soon.

**Comment by vsoch** [comment]
okay got something working (at least seemingly!)

image

I find this autoscaler thing still terrifying - mostly because it seems like it could explode at any point, although I suspect (hope) our min and max constraints are honored. There was almost no documentation on the Jupyter site for this - what I wound up doing is reading (again) here https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md and then seeing the deployment entrypoint that targets a specific label, one that we have defined in eksctl. Then it started to make sense. Important attributes for us to set are:

- min/max values (set for each node group in the eksctl config)
- max concurrent nodes (ditto, I currently have it fairly small, because, fear :scream: )
- hub->concurretSpawnLimit
- scheduler->userPlaceholder-replicas: I found and I think can help with the "waiting around for a pod" thing in that it creates faux users that can be evicted. https://z2jh.jupyter.org/en/latest/administrator/optimization.html#scaling-up-in-time-user-placeholders

The biggest source of error, I think, would likely come down to how disconnected these components seem. But if we do enough test runs and understand what is going on, I think we can guard against that.  Here are the changes https://github.com/flux-framework/Tutorials/pull/17/commits/4b9141479667b8947c9c395c01f03a7b1971fc7c let me know if you want to do a full test run soon (e.g., to get the Flux maintainers feedback as you mentioned).

**Comment by vsoch** [comment]
Update: don't merge, I need to test adding dyad first!

**Comment by vsoch** [comment]
@milroy we're done! Dyad is added and tested, and (during that testing) the autoscaling seemed to work too. :ship: :rocket:

**Comment by milroy** [comment]
LGTM! Merging.

**Pull request merged**